### ISSUE TITLE
`preferLocal` option does not run dependencies of dependencies

### ISSUE THREAD

**Issue by AndyOGo** [sindresorhus/execa#169]
Dependencies of locally installed packages aren't found and thus cause **ENOENT** errors.

May I guess local node modules aren't resolved like [`require.resolve`](https://nodejs.org/api/modules.html#modules_require_resolve_request_options)?

It throws:
```sh
Error: spawn npmversion ENOENT
```

A workaround is to add the needed package as `peerDependencies`, so that the package is directly available at the consumer, like is done here https://github.com/axa-ch/generic-release/pull/24/files

**Comment by ehmicky** [comment]
@AndyOGo could you please provide steps to reproduce the issue?

**Comment by ehmicky** [comment]
I think I now understand your problem, let me rephrase it: `execa` `preferLocal` option does not allow running nested dependencies (dependencies of dependencies). 

I don't think this is an issue. The way `npm` works (and most Node libraries) requires explicit declaration of dependencies to use them (either programatically or through the command line). This also seems like a good thing, as you are being explicit about a CLI that your program is using. 

`peerDependencies` is not a workaround, it's the actual solution for this kind of problem.

@sindresorhus can we close this issue?

**Comment by AndyOGo** [comment]
@ehmicky 
@sindresorhus 
Thanks for your answer.

Right, it's about nested dependencies.

To be clear we release a node package called `@axa-ch/generic-release` which has `npmversion` as it's own dependency. And this dependency is not found if our binary is execute in the scope of another project.

So it's about the dependencies of the dependent node package.
`execa` should be able to resolve these properly, please reopen this issue.

Node provides the perfect API for this case already by [`require.resolve`](https://nodejs.org/api/modules.html#modules_require_resolve_request_options).

Please let me make the point clear. that `require.resolve` is the proper solution, not `peerDependencies`.

**Comment by ehmicky** [comment]
Ok I think I see what you're trying to do with `@axa-ch/generic-release`. I actually have my release logic abstracted to [its own module](https://github.com/ehmicky/gulp-shared-tasks/blob/master/src/tasks/release/main.js) as well, so I understand where you're coming from.

Now I must be missing something because `node_modules/.bin/` (which is where `preferLocal` will find your binary) would contain the `npmversion` binary under your setup, i.e. everything should work.

For example, let's say:
  - you have a package called `@axa-ch/example`
  - `@axa-ch/example` has a `devDependency` called `@axa-ch/generic-release`
  - `@axa-ch/generic-release` has a production `dependency` called `npmversion`
  - both `@axa-ch/generic-release` and `npmversion` have binaries (`bin` field in `package.json`)

Under that setup, when `@axa-ch/example` call `npm install`, both binaries should be available under `node_modules/.bin/`, i.e. executable with `execa`. Let me know what I'm missing here.

**Comment by AndyOGo** [comment]
@ehmicky Thanks for your quick answer.

I agree, and I also thought that `node_modules/.bin` would be resolved of any nested deps, cause npm creates a flat directory structure. Unfortunately turned out my understanding of this was wrong.

We use `execa` because we needed a cross-platform way to automatically create SEMVER valid versions, run some `git`, `npm` and `npmversion` commands.

Indeed, you are right, would also be possible to call `npmversion` programatically.

Nevertheless this is an issue for all nested npm packages not being resolved by `execa`.

**Comment by ehmicky** [comment]
> I agree, and I also thought that node_modules/.bin would be resolved of any nested deps, cause npm creates a flat directory structure. Unfortunately turned out my understanding of this was wrong.

Could you share your setup? I am still not understanding how you're having issue there.

**Comment by AndyOGo** [comment]
I quickly revised my understanding.

As per NPM docs seems that only locally installed dependencies are available for [`npm run scrips`](https://docs.npmjs.com/cli/run-script#description):
> In addition to the shell’s pre-existing **PATH**, **npm run** adds `node_modules/.bin` to the **PATH** provided to scripts. Any binaries provided by **locally-installed dependencies** can be used without the `node_modules/.bin` prefix. For example, if there is a **devDependency** on **tap** in your package, you should write:
> ```sh
> "scripts": {"test": "tap test/\*.js"}
> ```

---

Here is one of our public setups:
1. [`@axa-ch/patterns-library`](https://github.com/axa-ch/patterns-library) (a collection of web-components) has [`@axa-ch/generic-release`](https://github.com/axa-ch/generic-release) in it's `devDependencies`:
   https://github.com/axa-ch/patterns-library/blob/c2ecb5a186903aa607291794cb78495a86737ebc/package.json#L83
2. and use it's binary within the npm scrips for releasing:
   https://github.com/axa-ch/patterns-library/blob/c2ecb5a186903aa607291794cb78495a86737ebc/package.json#L44
3. [`@axa-ch/generic-release`](https://github.com/axa-ch/generic-release) has `execa` and `npmversion` within it's `dependencies`:
   https://github.com/axa-ch/generic-release/blob/8e15bfde4632353ba9a4b8786dcae18f9df7fd5e/package.json#L18-L19

The issue is that executing `@axa-ch/generic-release` within the scope of `@axa-ch/patterns-library` `execa` doesn't find `npmversion` if it's not locally installed in that scope - so I added it as `peerDependencies` too:
https://github.com/axa-ch/generic-release/blob/8e15bfde4632353ba9a4b8786dcae18f9df7fd5e/package.json#L23-L25

**Comment by ehmicky** [comment]
npm docs says that `locally-installed dependencies can be used without the node_modules/bin/ prefix` but it does not say that nested dependencies cannot be used. As a matter of fact they can. If you check [`npm` source code](https://github.com/npm/npm-lifecycle/blob/latest/index.js#L108), they basically do the same as `execa` (add `node_modules/.bin/` to `$PATH`).

The setup you provided above does not produce any issues for me. Reproducible steps:

```bash
$ git clone https://github.com/axa-ch/patterns-library.git
$ cd patterns-library
$ npm install
$ vim node_modules/@axa-ch/generic-release/lib-release.js
$ npm run release
> @axa-ch/patterns-library@2.0.1-beta.264 release /home/ehmicky/patterns-library
> generic-release docs dist

Node-Version 1.7.0

A command line node module to deal with "bumping" and "npm version"
It should be execute into a folder with a package.json file

Usage: version [options]

Options:
    --help
        Print the help around the command

    -i --increment [<level>]
        Increment a version by the specified level.  Level can
        be one of: major, minor, patch, premajor, preminor,
        prepatch, or prerelease.  Default level is 'patch'.
        Only one version may be specified. A Git commit and
        tag will be created.
        Nota Bene: it will use the "npm version" command if the option
        "read-only" is not activated.

        -p --preid <identifier>
            Identifier to be used to prefix premajor, preminor,
            prepatch or prerelease version increments. It could
            be 'snapshot', 'beta' or 'alpha' for example.

        --force-preid
            If specified, we force to add if needed the specified preid

        --read-only
                Print only the future version. Don't modify the package.json file,
                nor the npm-shrinkwrap.json file, don't create a commit and don't
                create a git tag

        --nogit-commit
            No git commit

        --nogit-tag
            No git tag

        --git-create-branch
            Create a new branch. Does not work if a preid is generated.

        --git-push
            Push the commit, the branch and the tags if needed

    -u  --unpreid
        Remove the prefix. The increment and preid option will be ignored.
        Only a Git commit will be created

        --read-only
               Print only the future version. Don't modify the package.json file,
               nor the npm-shrinkwrap.json file, don't create a commit and don't
               create a git tag

       --nogit-commit
           No git commit

       --nogit-tag
           No git tag

        --git-create-branch
            Create a new branch. Does not work if a preid is generated.

        --git-push
            Push the commit, the branch and the tags if needed


🚀  Hello Dear developer, welcome to the release assistant. 🚀

!! Please make sure you have no changes to be commited !!

I'm getting some information....

>>> resolved | npm whoami
>>> resolved | npm owner ls
Attention: Your account ehmicky has no publisher rights. Please contact the administrator
```

In the `vim` command, I added the following line:

```diff
#!/usr/bin/env node

const readline = require('readline');
const outdent = require('outdent');
const execa = require('execa');
+ console.log(execa.sync('npmversion').stdout)
```

I.e. `execa` does find `npmversion` within `generic-release`. 

I don't think there is an issue with `execa` there. If you can get a reproducible setup which produces the bug, we can look into it.

**Comment by AndyOGo** [comment]
@ehmicky 

Thank you so much for looking into this and investing your valuable time.

Please watch out, as I said I declared `npmversion` as `peerDependencies`.
So I also added `npmversion` to the `devDependencies` of `@axa-ch/patterns-library`, as can be seen here:
https://github.com/axa-ch/patterns-library/blob/develop/package.json#L121

Without that it would not work.

**Comment by ehmicky** [comment]
I am surprised by this. `peerDependencies` does not actually make `npm` install anything. It only adds a warning when the parent package does not declare that dependency.

**Comment by ehmicky** [comment]
If you remove `npmversion` from the `devDependencies` of `@axa-ch/patterns-library` (and from the `peerDependencies` of `@axa-ch/generic-release`), the bug still does not appear. I just checked it.

**Comment by AndyOGo** [comment]
Interesting, I observer other behaviour.

Which OS are you using? We need to support Windows, Linux and macOS

**Comment by ehmicky** [comment]
I am on Linux. OS difference might be a thing, because `npm` install things quite differently on Windows. Are you on Windows?

**Comment by AndyOGo** [comment]
I'm on macOS, most people are on Windows

**Comment by ehmicky** [comment]
Alright if you're on macOS, there should be no difference in how `npm` installs things.
